xbmc/input/InputManager.h

```diff
@@ -353,6 +361,7 @@ class CInputManager : public ISettingCallback,
 
   std::vector<KODI::KEYBOARD::IKeyboardDriverHandler*> m_keyboardHandlers;
   std::vector<KODI::MOUSE::IMouseDriverHandler*> m_mouseHandlers;
+  std::vector<std::unique_ptr<IInput>> m_inputs;
```

Comment:
I didn't get the idea why manager should hold pointer to an input. They are independent now. in my opinion a creator should hold a pointer and manage an input lifetime.

Reply:
I agree, see comments above: Lirc.h include....

Reply:
btw. almost independent. manager sill holds the config for the mapping, i.e. Lircmap.xml

Reply:
no, they are absolutely independent. mappings may be different with the same implementation. I agree that using Lirc without mapping has no sense but in my view this is very weak reference.

Reply:
as an example Xbox remote control has no mappings at all

Reply:
If IInput depends on InputManager, then it's not safe to destruct it from IInputManager's destructor. It needs a lifetime outside IInputManager so it can be unregistered on shutdown.

Reply:
Lirc can be destructed at any time

Reply:
Not after InputManager, or `CServiceBroker::GetInputManager().TranslateLircRemoteString()` will segfault

Reply:
InputManger needs to register with a pointer at ServiceBroker.

Reply:
Good point, that would work. However, I've been wondering for a while, how do we prevent race conditions between when the pointer is accessed via `CServiceBroker::GetInputManager()` and when it is used? Return a shared pointer?

Reply:
Yes, shared ptr shoul work

Reply:
Yes, shared ptr shoul work

Reply:
But shared_ptr destroys init/deinit order. If InputManager has a dangling reference, it could be destroyed after something that it depends on is destroyed. We had this same problem with CPeripherals before.

Reply:
In final stage the order of init/deinit should not matter.

Reply:
how do we get from here to there?

Reply:
See this pr. Now lirc is independent from order of init. AE is next, only needs a little change. Powermanager shoulndˋt be a big deal either. Why donˋt you fix peripherials or some other component?